src/data_source.rs

```diff
@@ -13,4 +13,19 @@ use crate::BidiClass;
 /// character
 pub trait BidiDataSource {
     fn bidi_class(&self, c: char) -> BidiClass;
+    /// If this character is a bracket according to BidiBrackets.txt,
+    /// return the corresponding *normalized* *opening bracket* of the pair,
+    /// and whether or not it itself is an opening bracket.
+    ///
+    /// This effectively buckets brackets into equivalence classes keyed on the
+    /// normalized opening bracket.
+    ///
+    /// The default implementation will pull in a small amount of hardcoded data,
+    /// regardless of the `hardcoded-data` feature. This is in part for convenience
+    /// (since this data is small and changes less often), and in part so that this method can be
+    /// added without needing a breaking version bump.
+    /// Override this method in your custom data source to prevent the use of hardcoded data.
+    fn bidi_matched_opening_bracket(&self, c: char) -> Option<(char, bool)> {
```

Comment:
I'm not a fan of this API. Perhaps we should split this into two methods, one which handles bidi matching, and one which gets you normalization, ideally expected to only work for properties that also have bidi matched pair values set?

cc @sffc and @echeran since we'll have to deal with this upstream in ICU4X

Reply:
@sffc in part, but it also needs normalization (currently for a single character pair). Note that the mirrored_glyph property has two values, so the tuple API is still necessary.

As I said it could be split into two to move normalization out.

Reply:
Currently we only need to carry Bidi_Class data. I'm concerned that if we need both General_Category and Bidi_Mirrored and some normalization stuff, it will dramatically increase the size of our BiDi-only ICU4X build, even though the data seems to resolve down to only a small amount at the end of the day.

Perhaps ICU4X should resolve all of this at datagen time and introduce a new small key specifically for this purpose? What would that look like?

Reply:
FWIW, when I made my ICU4X-Bidi wasm build with this PR yesterday (the last commit at that time was `Handle canonical equivalence`), the size of the wasm build increased from ~20KB to ~23KB (brotli-compressed).

Not a deal breaker for us, but I won't be opposed to reducing the size if there are some low hanging fruits 🙂

Reply:
> Note that the trait model and the ICU4X data model are two separate questions, each trait model can be implemented with either data model.

I think I lean toward the fully resolved property ("frankenproperty") for the API, but it should return a struct instead of a `Option<(char, bool)>` to be more clear about what it is doing. It's nice that the API enforces that it only needs to be defined for fewer characters, rather than having `.normalize(char) -> Option<char>` which makes it look very tempting to be hooked up to a full normalizer, which is neither necessary nor efficient.

Reply:
Yeah of course. I can't approve PRs on this repo anyway.